### PR TITLE
Support capturing enum values and challenges

### PR DESCRIPTION
**Pull request by chriseth**
After this, we can process prover functions that capture enum values and challenges. This is needed to improve the lookup transformation functions.

### CODE REVIEW

File: pil-analyzer/src/condenser.rs [hunk 1]
```diff
@@ -940,16 +940,32 @@ fn try_value_to_expression<T: FieldElement>(value: &Value<'_, T>) -> Result<Expr
         }
         .into(),
         Value::Closure(c) => try_closure_to_expression(c)?,
-        Value::TypeConstructor(c) => {
+        Value::TypeConstructor(enum_, c) => {
             return Err(EvalError::TypeError(format!(
-                "Type constructor as captured value not supported: {c}."
+                "Type constructor as captured value not supported: {}::{c}.",
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply pil-analyzer/src/condenser.rs:945]: Related to other comment about using structs: it would remove this occurence of `::` by calling a method on the struct.

File: pil-analyzer/src/condenser.rs [hunk 2]
```diff
@@ -940,16 +940,32 @@ fn try_value_to_expression<T: FieldElement>(value: &Value<'_, T>) -> Result<Expr
...
+            );
+            match items {
+                None => variant_ref,
+                Some(items) => FunctionCall {
+                    function: Box::new(variant_ref),
+                    arguments: items
+                        .iter()
+                        .map(|i| try_value_to_expression(i))
+                        .collect::<Result<_, _>>()?,
+                }
```
Comment (Schaeff) [inline pil-analyzer/src/condenser.rs:966]: Not directly related to this PR, but couldn't we simplify this by making the items non optional, and making `MyEnum::NoArgs` sugar for `MyEnum::NoArgs()`?

File: pipeline/src/lib.rs [hunk 3]
```diff
@@ -77,6 +77,7 @@ impl HostContext {
 // TODO at some point, we could also just pass evaluator::Values around - would be much faster.
 pub fn parse_query(query: &str) -> Result<(&str, Vec<&str>), String> {
     // We are expecting an enum value
+    let query = query.strip_prefix("std::prelude::Query::").unwrap_or(query);
```
Comment (Schaeff) [inline pipeline/src/lib.rs:80]: Is this change needed, and is so, why?

File: pil-analyzer/src/condenser.rs [hunk 4]
```diff
@@ -626,15 +626,15 @@ fn to_constraint<T: FieldElement>(
     counters: &mut Counters,
 ) -> AnalyzedIdentity<T> {
     match constraint {
-        Value::Enum("Identity", Some(fields)) => {
+        Value::Enum(_, "Identity", Some(fields)) => {
```
[the remark this answers is not in the document]
Comment (chriseth) [reply pil-analyzer/src/condenser.rs:629]: True, now we can add an assertion to check what the enum is.

File: pil-analyzer/src/condenser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chriseth) [reply pil-analyzer/src/condenser.rs:945]: You mean the detail in the error string that this is about a "captured" value should move to the caller since this function could also be used for other values?

File: pil-analyzer/src/condenser.rs (hunk 2 above)
Comment (chriseth) [reply pil-analyzer/src/condenser.rs:966]: Just checking if I understand this correctly. Let's assume we have```
enum A {
  X,
  Y(),
}
```
(setting aside the question if we want to support an enum variant of the kind `Y()` - or was that the main question?), then `A::X` is an enum value while `A::Y` as a function that returns an enum value.

What we could do is change the items in `evaluator::Value::Enum` from `Option<Vec<Arc<Self>>>` to `Vec<Arc<Self>>`, since for a fixed enum variant, we do not need to distinguish between `None` and `Some(vec![])`.

Is that what you meant?

File: pipeline/src/lib.rs (hunk 3 above)
Comment (chriseth) [reply pipeline/src/lib.rs:80]: Because enum values are now printed properly with the enum name: https://github.com/powdr-labs/powdr/pull/1799/files#diff-f00565b83aa549438943c22a062d9bd3e6c27f77e39d2dcb6edbf456aa53e76aR226

I.e. before this, it would just print `Input`, now it prints `std::prelude::Query::Input`

File: pil-analyzer/src/condenser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Schaeff) [reply pil-analyzer/src/condenser.rs]: Yes.

File: pil-analyzer/src/condenser.rs (hunk 2 above)
Comment (Schaeff) [reply pil-analyzer/src/condenser.rs]: Getting rid of the `Option<Vec<_>>` would indeed be nice, but what I meant was:

Here if we have a variant with no arguments we return an enum value, otherwise we return a function call. I was thinking we could remove this edge case and always use `A::X()` internally to represent `A::X`. However it seems like it would be ambiguous because `A::X` would then be both a function and an enum value.

Closing

**Review by Schaeff — APPROVED** [review]

File: pil-analyzer/src/condenser.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chriseth) [reply pil-analyzer/src/condenser.rs]: -> https://github.com/powdr-labs/powdr/pull/1819
